subsys/pm/device.c

```diff
@@ -151,3 +151,37 @@ int pm_device_state_get(const struct device *dev,
 	return dev->pm_control(dev, PM_DEVICE_STATE_GET,
 			       device_power_state, NULL, NULL);
 }
+
+
+bool pm_device_wakeup_enable(struct device *dev, bool enable)
+{
+	atomic_val_t flags, new_flags;
+
+
+	flags =  atomic_get(&dev->pm->atomic_flags);
+
+	if ((flags & BIT(PM_DEVICE_ATOMIC_FLAGS_WS_CAPABLE_BIT)) == 0U) {
+		return false;
+	}
+
+	if (enable) {
+		new_flags = flags |
+			BIT(PM_DEVICE_ATOMIC_FLAGS_WS_ENABLED_BIT);
+	} else {
+		new_flags = flags & ~BIT(PM_DEVICE_ATOMIC_FLAGS_WS_ENABLED_BIT);
+	}
+
+	return atomic_cas(&dev->pm->atomic_flags, flags, new_flags);
+}
```

Comment:
I'd personally simplify this function:

1. Make it `void` return type
2. Do nothing if the device is not wakeup capable (and document that this call will have no effect on non-wakeup capable devices)

Summarizing:
```c
void pm_device_wakeup_enable(struct device *dev, bool enable)
{
        if (!atomic_test_bit(&dev->pm->atomic_flags, PM_DEVICE_ATOMIC_FLAGS_WS_CAPABLE_BIT)) {
                return;
        }

        atomic_set_bit_to(&dev->pm->atomic_flags, PM_DEVICE_ATOMIC_FLAGS_WS_ENABLED_BIT, enable);
}
```

Reply:
I don't. If two threads call it simultaneously, one enabling and other disabling how would you know the result ? Using compare and set allows the caller to know if it succeed.

Reply:
If two threads compete to do the opposite the last to do the operation will win, which I guess it's fine. Maybe I'm missing a usecase, though.

Reply:
it is fine  but  who called it would no be able to know if it succeed.  I rather have it available.

Reply:
If two threads are competing to do the opposite, it's an application design issue, not a failure. This API per se does not have any failure path when designed, e.g. with `atomic_set_bit_to`. If the application/device needs to check if the device is really enabled or not, it should call, e.g. `pm_device_wakeup_is_enabled` later.

Reply:
the return of this function tells the caller whether or not it succeed also it returns false in case the device is not capable. We can eternally argue whether it is an application responsibility or not. I rather this way and it if it does not contain any flaw I would like to move with this implementation. Is it going to be a reason to block this pr ?

Reply:
The problem is that `pm_device_wakeup_enable` can't fail as it is designed.

In case of Linux, for example, it can:
https://elixir.bootlin.com/linux/latest/source/drivers/base/power/wakeup.c#L336

If it needs to return a failure, then it's return type should be an `int` and return a negative errno. For example, if the device is not wakeup capable it could return `-EINVAL` as Linux does.

Reply:
It can fail returning false. That means that operation did not succeed. Though, using an integer allows us to know what was the cause. Will change to that since you are pushing back for returning a boolean. But again, return false is ok for failing it tells that the operation didn't work.

Reply:
btw, this kind of implies in changing `bool pm_device_wakeup_enable` to return `int` since it can fail because the device is not wake up capable. This api was just return false in case of failure and true in success. Is that really the path you wanna go ?

Reply:
So how can a flag set fail? This is the part I don't really get.

Reply:
fail if the device is not capable. You cannot enable a wakeup source if the device does no support it.

Reply:
the way I'm using `atomi_cas` just ensure that no one else has changed this flags while this function is called, as for example if the device somehow loses the wake up capability. Of course, we are not covering it and just assuming that the never changes it is wakeup capability. So, use `atomic_set_bit_to` works but I tried to ensure that while I'm setting this flag no one else change these flags.